src/control/lib/control/pool.go

```diff
@@ -1057,89 +1063,212 @@ func processSCMSpaceStats(log debugLogger, filterRank filterRankFn, scmNamespace
 
-	resp, err := StorageScan(ctx, rpcClient, &StorageScanReq{Usage: true})
-	if err != nil {
-		return 0, 0, err
+	// Request parameters used to query and calculate space requirements for pool.
+	spaceReq := processSpaceReq{
+		log:               rpcClient,
+		filterRank:        newFilterRankFunc(ranks),
+		rankNVMeFreeSpace: make(rankFreeSpaceMap),
 	}
+	ramAvailBytes := uint64(math.MaxUint64)
 
-	if len(resp.HostStorage) == 0 {
-		return 0, 0, errors.New("Empty host storage response from StorageScan")
+	// Retrieve lowest RAM-disk available capacity across all pool ranks.
+	for _, key := range scanResp.HostStorage.Keys() {
+		hostStorage := scanResp.HostStorage[key].HostStorage
+
+		if hostStorage.ScmNamespaces.Usable() == 0 {
+			return 0, 0, errors.Errorf("host without ram storage: hostname=%s",
+				scanResp.HostStorage[key].HostSet.String())
+		}
+
+		usable, err := processSCMSpaceStats(spaceReq, hostStorage.ScmNamespaces)
+		if err != nil {
+			return 0, 0, err
+		}
+		rpcClient.Debugf("ram-disk: %+v, usable: %s", hostStorage.ScmNamespaces,
+			humanize.Bytes(usable))
+
+		if ramAvailBytes > usable {
+			ramAvailBytes = usable
+		}
+	}
+
+	if ramAvailBytes == math.MaxUint64 {
+		return 0, 0, errors.Errorf("no usable ram space found with rank list %s", ranks)
 	}
 
-	// Generate function to verify a rank is in the provided rank slice.
-	filterRank := newFilterRankFunc(ranks)
-	rankNVMeFreeSpace := make(rankFreeSpaceMap)
+	metaDataRolesSharedSet := false
+	metaDataRolesShared := false
+
+	for _, key := range scanResp.HostStorage.Keys() {
+		hostStorage := scanResp.HostStorage[key].HostStorage
+
+		isShared := areMetaDataRolesShared(hostStorage.NvmeDevices)
+		if metaDataRolesSharedSet && metaDataRolesShared != isShared {
+			return 0, 0, errors.Errorf("md-on-ssd role configs differ between hosts")
+		}
+		metaDataRolesShared = isShared
+		metaDataRolesSharedSet = true
+
+		if err := processNVMeSpaceStats(spaceReq, hostStorage.NvmeDevices); err != nil {
+			return 0, 0, err
+		}
+	}
+
+	// Retrieve lowest capacity available for DATA-on-SSD across all pool ranks.
+	dataBytes := uint64(math.MaxUint64)
+	for _, nvmeRankBytes := range spaceReq.rankNVMeFreeSpace {
+		if dataBytes > nvmeRankBytes {
+			dataBytes = nvmeRankBytes
+		}
+	}
+
+	// metaBytes indicates the capacity to be reserved for metadata across SSDs with META role.
+	metaBytes := ramAvailBytes
+	if createReq.MemRatio > 0 {
+		metaBytes = uint64(float64(metaBytes) / float64(createReq.MemRatio))
+	}
+
+	// In the case that META+DATA roles are both assigned to the same bdev_tier, the available
+	// capacity for DATA should be calculated as the total available capacity on all tier SSDs
+	// LESS metaBytes.
+	var msgMeta string
+	if metaDataRolesShared {
+		msgMeta += fmt.Sprintf(" with %s of reserved metadata capacity",
+			humanize.Bytes(metaBytes))
+		if metaBytes > dataBytes {
+			return 0, 0, errors.Errorf("insufficient free nvme space%s", msgMeta)
+		}
+		dataBytes -= metaBytes
```

Comment:
I've been discussing with @knard-intel about how we are doing NVMe size adjustments in both client-side control-API `GetMaxPoolSize()` and server-side control-plane `adjustNvmeSize()` and `getMetaClusterCount()`. It would be better to only perform the adjustments in one place and the next question is whether we can perform all the necessary adjustments on the server side.

Deriving the canonical per-rank pool sizes requires the following information:
- lowest total free SSD capacity across all pool ranks
- lowest ramdisk available capacity across all pool ranks
- mem-ratio
- number of SSDs in tiers
- roles of tiers

Whilst some of this information can be gathered within a single control-plane instance and some can be passed in via the scan request, the information that needs to be aggregated across hosts needs to be evaluated at the control-API level. With this in mind it seems sensible to leave these adjustments in the control API layer.

A subsequent related concern is that we are now reserving more META capacity than necessary in situations where META+DATA roles are shared. We reserve capacity both in `getMaxPoolSizeMdOnSsd()` and in `getMetaClusterCount()`.

The meta_size adjustments using `SmdDevice.MetaSize` in `getMetaClusterCount()` are currently hardcoded to remove 6gib*tgtCount from capacity if META role is set. This needs addressing because:
- mem-ratio may not be 100%
- meta-blobs may be spread across multiple META-role SSDs (in which case not all targets will be allocated to the same SSD)

What scan use cases do we have for making these meta_size adjustments? Currently I see:
- dmg storage query usage (where we are trying to evaluate capacity used by pool allocations and free space for new allocations)
- dmg pool create --size % (where we are trying to calculate available capacity to autosize new pool)
- dmg pool list -v (similar to query usage above)

Are there any other situations that I have missed?

Reply:
Another approach, calculate usage individually for each rank then choose the lowest of all.

meta_sz calculated from ns.Usable() with mem_ratio passed into metaRdbComputeSz()
but it still needs to be adjusted based on the number of targets per ssd

proposed draft https://github.com/daos-stack/daos/pull/15202